src/main/java/donna/Donna.java

```diff
@@ -0,0 +1,180 @@
+     * parser, and loading previously saved tasks.
+     */
+    public Donna() {
+        ui = new Ui();
+        storage = new Storage(DIRECTORY_PATH, FILE_PATH);
+        parser = new Parser();
+        try {
+            tasks = new TaskList(storage.loadTasks());
+        } catch (DonnaException e) {
+            ui.printErrorMessage(e.getMessage());
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Runs the main loop, processing user input and executing commands.
+     * Handles exceptions and prints error messages as needed.
+     */
+    private void run() {
+        boolean dataWasLoaded = !tasks.isEmpty();
+        ui.printGreeting(dataWasLoaded);
+
+        Scanner sc = new Scanner(System.in);
+        while (true) {
+            String input = sc.nextLine();
+            try {
+                ParsedCommand result = parser.parse(input);
+                String commandType = result.getCommandType();
+
+                switch (commandType) {
+                case "exit":
+                    ui.printGoodbyeMessage();
+                    storage.saveTasks(tasks);
+                    sc.close();
+                    return;
+                case "list":
+                    ui.printTaskList(tasks);
+                    break;
+                case "mark":
+                    handleMark(result.getArgument1());
+                    break;
+                case "unmark":
+                    handleUnmark(result.getArgument1());
+                    break;
+                case "delete":
+                    handleDelete(result.getArgument1());
+                    break;
+                case "add":
+                    handleAdd(result.getArgument1(), result.getArgument2());
+                    break;
+                case "find":
+                    List<Task> foundTasks = tasks.searchTasks(result.getArgument1());
+                    ui.printFindResults(foundTasks);
+                }
+            } catch (DonnaException e) {
+                ui.printErrorMessage(e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Handles the marking of a task as done.
+     *
+     * @param taskNum S.No of the task to mark (index from 1).
+     * @throws DonnaException If the argument is not a valid task number.
+     */
+    private void handleMark(String taskNum) throws DonnaException {
+        try {
+            int taskIdx = Integer.parseInt(taskNum) - 1;
+            Task task = tasks.markTask(taskIdx, true);
+            ui.printTaskMarkedMessage(task, true);
+        } catch (NumberFormatException e) {
+            ui.printErrorMessage(DonnaException.invalidTaskNumber().getMessage());
+        }
+    }
+
+    /**
+     * Handles the marking of a task as not done.
+     *
+     * @param taskNum Index of the task to unmark (index from 1).
+     * @throws DonnaException If the argument is not a valid task number.
+     */
+    private void handleUnmark(String taskNum) throws DonnaException {
```

Comment:
Since handleUnmark, handleAdd etc are all the same type of methods, it might be better to put them in a separate class of their own, or have different types of Commands and ways of executing them.

Reply:
That's a valid point. I'll do so.